codelabs/custom_renderer.md

```diff
@@ -0,0 +1,409 @@
+author: Rachel Fenichel
+summary: Codelab to create a custom renderer
+id: custom-renderer
+categories: codelab,blockly,rendering,customization
+status: Draft
+Feedback Link: https://github.com/google/blockly-samples/issues/new
+
+# Custom renderers
+
+## Codelab overview
+
+### What you'll learn
+In this codelab you will learn:
+1. How to define and register a custom renderer.
+1. How to override renderer constants.
+1. How to change the shape of connection notches.
+1. How to set a connection's shape based on its type checks.
+
+### What you'll build
+
+Over the course of this codelab you will build and use four renderers.
+
+1. A minimal custom renderer
+![](assets/custom-renderer/custom_renderer.png)
+1. A renderer with custom constants
+![](assets/custom-renderer/custom_constants.png)
+1. A renderer with custom notches
+![](assets/custom-renderer/custom_notches.png)
+1. A renderer with connection shapes that depend on the type checks of each connection.
+![](assets/custom-renderer/typed_connection_shapes.png)
+### What you'll need
+This codelab assumes that you are already comfortable with using the Blockly playground locally.
+
+## Setup
+
+In this codelab you will add code to the Blockly playground to create and use a new renderer.
+
+When the instructions say to "create and include a new file", you should place that file in the same folder as the playground and include it with a script tag.
+
+```
+<script src="custom_renderer.js"></script>
+```
+
+Note: you must include your custom code *after* including the Blockly library.
+
+## Minimal custom renderer
+
+This is the minimum code needed to define a new renderer.
+
+Create and include a new file named `custom_renderer.js`.
```

Comment:
I would pick either ### Define your renderer and ###Define your constant provider as separate headers or ### Define your renderer and constants provider as a single header and then use across all sections.